core/templates/pages/admin-page/features-tab/admin-features-tab.component.ts

```diff
@@ -0,0 +1,249 @@
+    private adminTaskManager: AdminTaskManagerService,
+    private apiService: PlatformFeatureAdminBackendApiService,
+    private ruleFactory: PlatformParameterRuleObjectFactory,
+    private filterFactory: PlatformParameterFilterObjectFactory,
+  ) {}
+
+  async reloadFeatureFlagsAsync(): Promise<void> {
+    const data = await this.adminDataService.getDataAsync();
+
+    this.featureFlags = data.featureFlags;
+
+    this.featureFlagNameToBackupMap = new Map(
+      this.featureFlags.map(feature => [feature.name, cloneDeep(feature)]));
+  }
+
+  addNewRuleToTop(feature: PlatformParameter): void {
+    feature.rules.unshift(
+      this.ruleFactory.createFromBackendDict({
+        filters: [{
+          type: PlatformParameterFilterType.ServerMode,
+          conditions: [['=', ServerMode.Dev.toString()]]
+        }],
+        value_when_matched: false
+      })
+    );
+  }
+
+  addNewRuleToBottom(feature: PlatformParameter): void {
+    feature.rules.push(
+      this.ruleFactory.createFromBackendDict({
+        filters: [{
+          type: PlatformParameterFilterType.ServerMode,
+          conditions: [['=', ServerMode.Dev.toString()]]
+        }],
+        value_when_matched: false
+      })
+    );
+  }
+
+  addNewFilter(rule: PlatformParameterRule): void {
+    rule.filters.push(
+      this.filterFactory.createFromBackendDict({
+        type: PlatformParameterFilterType.ServerMode,
+        conditions: []
+      })
+    );
+  }
+
+  addNewCondition(filter: PlatformParameterFilter): void {
+    const context = this.filterTypeToContext[filter.type];
+    filter.conditions.push([
+      context.operators[0],
+      context.options ? context.options[0] : ''
+    ]);
+  }
+
+  removeRule(feature: PlatformParameter, ruleIndex: number): void {
+    feature.rules.splice(ruleIndex, 1);
+  }
+
+  removeFilter(rule: PlatformParameterRule, filterIndex: number): void {
+    rule.filters.splice(filterIndex, 1);
+  }
+
+  removeCondition(
+      filter: PlatformParameterFilter, conditionIndex: number): void {
+    filter.conditions.splice(conditionIndex, 1);
+  }
+
+  moveRuleUp(feature: PlatformParameter, ruleIndex: number): void {
+    const rule = feature.rules[ruleIndex];
+    this.removeRule(feature, ruleIndex);
+    feature.rules.splice(ruleIndex - 1, 0, rule);
+  }
+
+  moveRuleDown(feature: PlatformParameter, ruleIndex: number): void {
+    const rule = feature.rules[ruleIndex];
+    this.removeRule(feature, ruleIndex);
+    feature.rules.splice(ruleIndex + 1, 0, rule);
+  }
+
+  async updateFeatureRulesAsync(feature: PlatformParameter): Promise<void> {
+    if (this.adminTaskManager.isTaskRunning()) {
+      return;
+    }
+    const commitMessage = this.windowRef.nativeWindow.prompt(
+      'This action is irreversible. If you insist to proceed, please enter ' +
+      'the commit message for the update',
+      `Update feature '${feature.name}'.`
+    );
+    if (commitMessage === null) {
+      return;
```

Comment:
Suggest setting a status message in this case rather than silently failing.

Reply:
Ah, to be clear I was considering the case when the user gets the confirmation dialog & doesn't fill in a commit message, then clicks confirm. Won't that look the same as cancelling with the current logic?